WalletWasabi/Wallets/WalletManager.cs

```diff
@@ -439,7 +439,15 @@ public bool AnyCoinJoinInProgress()
 		{
 			lock (Lock)
 			{
-				return Wallets.Any(x => x.ChaumianClient?.State.AnyCoinsQueued() is true);
+				return Wallets.Any(w => w.Coins.Unspent().CoinJoinInProcess().Any());
+			}
+		}
+
+		public bool AnyCoinJoinInCriticalPhase()
+		{
+			lock (Lock)
+			{
+				return Wallets.Any(w => w.Coins.Unspent().CoinJoinInCritical().Any());
```

Comment:
But this simply says that `CoinJoinInCritical` is not property of a coin but rather of some higher level entity, right? 

Could an alternative be that CoinjoinClient or CoinjoinManager simply calls a callback or fires an event?

Reply:
The logic is coming from ICoinsView. I applied the same implementation as we do with `CoinJoinInProcess`. According to the current logic of CoinJoinManager (which I am OK with), it will check the current status of the related objects and do the required actions. So the concept here is to avoid callbacks, events and just present the required information. The simplest solution I can think of - no concurrently or deadlocks - there will be one state update with every iteration of CoinJoinManager and one object to make all decisions related to CoinJoin.

Reply:
I still think it would better to have:

* `SystemAwakeChecker` not being a periodic runner
* Have a callback in (probably) `CoinJoinClient` reporting when CJ is in critical state. To avoid deadlock here, one can even send a message (using an async queue or `Channel<MsgType>` or whatever to avoid deadlocks), with logging this is easy to debug.

Then 

* `SystemAwakeChecker` would not know about `CoinJoinClient` and vice versa (helps testing).
* Behavior of calling a callback can be easily tested.
* `SmartCoin` would stay more lean (I don't think it should know anything about higher level functionality as that can get out of hand in the future).

Reply:
Pushing is more efficient/clean/correct while pulling is easier to implement and at this moment we need something really easy to implement and understand.

I had to face a similar dilemma when @molnard and me independently implemented the same feature, I decided to go with the simpler one (David's one). This doesn't mean I agree with the `SmartCoin` should be use to pass status/messages across different components, no, i don't think it is a good idea but it is a debt we should take in order to buy time.

Reply:
Check this https://github.com/zkSNACKs/WalletWasabi/pull/6321